### PR TITLE
Replace better-eval with vm

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-lf**
Regarding issue 366

The PR replaces better-eval with vm for converting raw variant strings. The better-eval package creates a new VM for each conversion while this change creates a single VM and re-uses it for all conversions

### CODE REVIEW

**Comment by sfc-gh-ext-simba-lf** [comment]
Added test case similar to the repro that calls select on a variant column containing a large result set

**Review by sfc-gh-igarish — APPROVED** [review]
Please do pref test and put numbers in this ticket before and after the change.

**Comment by sfc-gh-ext-simba-lf** [comment]
> Please do pref test and put numbers in this ticket before and after the change.

Average execution time of testVariantLarge over 10 runs: 
 - better-eval: 22.6s
 - vm: 4.4s

File: package.json
Comment (sfc-gh-dprzybysz) [inline package.json:39]: Do we need to add this library? It was released 7 years ago https://www.npmjs.com/package/vm?activeTab=readme and it has information that we should not install it since it's part of node now https://github.com/DiegoRBaquero/node-vm
(I removed this dependency and it seems to work fine)

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:184]: if this assertion fails then callback in line below won't be called and the test could stuck till timeout?

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:180]: shouldn't we wrap next lines in `else`?

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:233]: the main focus of the test is to insert and select from table with variant - maybe we can wrap this test in describe and create table put in `before` and drop table in `after`?

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:220]: if this assertion fails than callback in line below won't be called and we will wait till timout?

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:124]: array could also be `const`

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:120]: We are sending one gigantic array in json and let Snowflake to parse it and store as one json per row. Why don't we send binds as array of arrays? is it important for the test?

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testLargeResultSet.js:117]: could we name this test in more elaborated way? to tell what is the core of the test? is it inserting or selecting?

File: package.json
Comment (sfc-gh-ext-simba-lf) [reply package.json]: It's removed now

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testLargeResultSet.js]: Sure, test is renamed now

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testLargeResultSet.js]: Done

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testLargeResultSet.js:216]: You're right, it would've called callback again otherwise. Added else clause

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testLargeResultSet.js]: It goes to the catch clause and calls callback(err)

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testLargeResultSet.js]: Ok, placed the test in its own describe and create/drop table are now in the before/after of the test

File: test/integration/testLargeResultSet.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testLargeResultSet.js]: This is how we were advised by the previous engineer when inserting multiple rows of variant data

I tried binding an array of arrays but got `OperationFailedError: SQL compilation error: error line 1 at position 53 Bind variable ? not set.`
```
sqlText: "insert into testVariantTable(colA) select parse_json(?)",
binds: [ [JSON.stringify({ 'root': 'value' })], [JSON.stringify({ 'root': 'value' })] ],
```

I then tried using insert into values but got a `Error [OperationFailedError]: SQL compilation error: Unsupported data type 'VARIANT'.` instead
```
sqlText: "insert into testVariantTable(colA) values(?)",
binds: [ [{ 'root': 'value' }], [{ 'root': 'value' }] ],
```

However, that error did not show up when array binding non-variant data
```
sqlText: 'insert into testTable values(?)',
binds: [[1], [2], [3]],
```

And running the first query without array bind was successful
```
sqlText: 'insert into testVariantTable(colA) select parse_json(?)',
binds: [JSON.stringify({ 'key': 'value' })],
```

Is there a different combination of sql and binds that I should be using instead?

File: lib/connection/result/column.js
Comment (yanavais) [reply lib/connection/result/column.js:557]: Additional input that might be helpful. We noticed that `better-eval` is very slow in JSON parsing comparing to JSON.parse(). Most, if not all of our `variant` fields are JSON. Maybe it makes sense to apply `JSON.parse() first and leave `better-eval` as a fallback.

File: test/integration/testDataType.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testDataType.js:34]: let's add new code with const instead of var

File: test/integration/testDataType.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testDataType.js:254]: can we create describe for variant and in before/afterEach/after add createTable/truncateTable/dropTable?

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_config_test.js:34]: instead of comment let's wrap test cases in describe

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_config_test.js:96]: let's create describe instead of comments

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_config_test.js:89]: we can simplify test cases to 
```
negativeTestCases.forEach(testCase => {it(testCase.name, createNegativeITCallback(testCase))})
```
or even inline createNegativeITCallback in test body

The same could be done with positive cases

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_config_test.js:175]: is `new` missing when XMLParser is created? here we give example where exception is probably thrown by who would catch it in our production code when variant is parsed?

**Review by sfc-gh-dprzybysz — CHANGES_REQUESTED** [review]
Let's add run lint (`npm run lint:check -- FILENAME`) for new code. Especially new files could be even changed with `npm run lint:fix -- FILENAME`.

File: test/integration/testDataType.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testDataType.js:220]: can we add additional describe and tests where we override default json and xml parser?

alternative implementation for JSON could be just `JSON.parse` and for xml we can leave todo to implement the test in https://github.com/snowflakedb/snowflake-connector-nodejs/pull/538 (e.g. `TODO SNOW-830291: add custom xml parser test` should be enough)

**Comment by odedpeer** [comment]
> > Please do pref test and put numbers in this ticket before and after the change.
> 
> Average execution time of testVariantLarge over 10 runs:
> 
> * better-eval: 22.6s
> * vm: 4.4s

what is the average execution time of `eval`? I'd like to see the difference between `vm` and `eval`

File: lib/connection/result/column.js
Comment (sfc-gh-ext-simba-lf) [reply lib/connection/result/column.js:564]: When logging with `%o`, the error stack shows the data that was being parsed. Should we remove that data before logging?

Currently, only the error message is logged

File: lib/connection/result/column.js
Comment (sfc-gh-ext-simba-lf) [reply lib/connection/result/column.js:559]: Oops sorry I didn't realize I deleted it. I saw that comment but I wasn't sure whether to keep the logging/catch inside the default xml parser function since the default parser could be replaced

I thought about wrapping it in its own try/catch so it doesn't rely on the custom function to do the logging/catching. Do you think a nested try/catch for the xml parsing is ok or keeping it in the default function is enough?

File: lib/global_config.js
Comment (sfc-gh-ext-simba-lf) [reply lib/global_config.js:14]: Done

File: lib/global_config.js
Comment (sfc-gh-ext-simba-lf) [reply lib/global_config.js]: Ok

File: lib/global_config.js
Comment (sfc-gh-ext-simba-lf) [reply lib/global_config.js:208]: Alright, added the else and throw

File: lib/global_config.js
Comment (sfc-gh-ext-simba-lf) [reply lib/global_config.js]: Added the expected signature. I put down `any` as the expected result, let me know if it should be `Object`

File: lib/global_config.js
Comment (sfc-gh-ext-simba-lf) [reply lib/global_config.js]: Added the expected signature

File: test/integration/testDataType.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testDataType.js]: Ok

File: lib/core.js
Comment (sfc-gh-ext-simba-lf) [reply lib/core.js:174]: Do you prefer closing that PR after this is merged? Or close it now

File: test/integration/testDataType.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testDataType.js]: Sure, I also copied the way you used `executeCmdAsync` from the large result test case

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_config_test.js]: Ok

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_config_test.js]: Ok

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_config_test.js]: Nice, I'll use this for both the test cases. I put the callback function's body inside if that's what you meant

File: test/unit/snowflake_config_test.js
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_config_test.js]: You're right, it was missing. Added `new` when creating XMLParser and corrected the parenthesis placement

I mentioned in a comment above about wrapping the xml parsing in its own try/catch if possible

File: test/integration/testDataType.js
Comment (sfc-gh-ext-simba-lf) [reply test/integration/testDataType.js]: I tried using `JSON.parse()` but it failed because the test data contained `Infinity` and `undefined` which doesn't seem to be supported by JSON

I made another insert query with only valid JSON and it works

File: lib/global_config.js
Comment (sfc-gh-pmotacki) [inline lib/global_config.js:216]: In changes of SNOW-830291 I've prepared possibility to set custom xmlParserConfiguration:  
/**
 * Set custom xmlParser configuration.
 * Parametrs that you can override:
 *   ignoreAttributes - default true,
 *   attributeNamePrefix - default '@',
 *   attributesGroupName - default null,
 *   alwaysCreateTextNode - default false
 *
 * @param {object} value
 */
Examle:
```
snowflake.configure({
    xmlParserConfiguration: {
        ignoreAttributes: true,
        attributeNamePrefix: '@'
    },
})
``` 
and it is used during creation of new XMLParser
I wonder if it will be useful to have both ways configured? This is more to discuss with @sfc-gh-igarish. When we decide I will implement it in my task

File: lib/core.js
Comment (sfc-gh-dprzybysz) [reply lib/core.js:174]: Definitely after with comment like "It is fixed in ..."

File: lib/connection/result/column.js
Comment (sfc-gh-dprzybysz) [reply lib/connection/result/column.js:559]: it's hard question... but maybe we should keep it consistent with json parser and log also here

Maybe we should have such logic here (is pseudo js):

```
try {
  return parseJson()
} catch (jsonParseError) {
  try {
     return parseXml()
  } catch(xmlParseError) {
    log(variant cannot be parsed neither as JSON (jsonParseError.message) nor as XML (xmlParseError.message)
    throw new VariantParseError(jsonParseError, xmlParseError)
  }
}
```

@sfc-gh-igarish WDYT?

File: lib/global_config.js
Comment (sfc-gh-dprzybysz) [reply lib/global_config.js]: It's interesting question - if we have sth declared as variant then we can put as json `{}`, `[]` or `5` and all of them are  considered by parser as valid jsons? (https://www.json.org/json-en.html)

Comment (sfc-gh-ext-simba-lf) [reply lib/global_config.js]: It looks like they are considered valid, no errors were thrown by the default json parser when parsing those 3 values

File: lib/connection/result/column.js
Comment (sfc-gh-ext-simba-lf) [reply lib/connection/result/column.js:559]: This could work, info on both error messages is provided through the logging and VariantParseError

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: lib/connection/result/column.js
Comment (sfc-gh-ext-simba-lf) [reply lib/connection/result/column.js]: Hello, apologies for the delay, the decision to not use `eval` was based on the security team's advise. The parser can be configured now though so if there is a parser you'd like to use over `vm`, then it is possible to configure:

- Configure the connector to use `eval` when parsing json

```
snowflake.configure({
  jsonColumnVariantParser: rawColumnValue => eval("(" + rawColumnValue + ")");
})
```
- Or use `JSON.parse()` instead
```
snowflake.configure({
  jsonColumnVariantParser: rawColumnValue => JSON.parse(rawColumnValue)
})
```

**Pull request merged**
